pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,252 @@
+	RuleProtoAll
+	RuleProtoTCP
+	RuleProtoUDP
+	RuleProtoICMP
+)
+
+type MssClampType uint32
+
+const (
+	UndefinedMSS MssClampType = iota
+	ToPMTU
+	ToValue
+)
+
+type ChainHook uint32
+
+const (
+	ChainHookPrerouting ChainHook = iota
+	ChainHookInput
+	ChainHookForward
+	ChainHookOutput
+	ChainHookPostrouting
+	ChainHookMAX
+)
+
+type ChainPriority uint32
+
+const (
+	ChainPriorityFirst ChainPriority = iota
+	ChainPriorityLast
+)
+
+type ChainType uint32
+
+const (
+	ChainTypeFilter ChainType = iota
+	ChainTypeRoute            // mangle
+	ChainTypeNAT
+	ChainTypeMAX
+)
+
+type ChainPolicy uint32
+
+const (
+	ChainPolicyAccept ChainType = iota
+	ChainPolicyDrop
+)
+
+type Rule struct {
+	DestCIDR string
+	SrcCIDR  string
+
+	SrcSetName  string
+	DestSetName string
+
+	SnatCIDR string
+	DnatCIDR string
+
+	OutInterface string
+	InInterface  string
+	TargetChain  string
+	MssValue     string
+
+	DPort     string
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type ChainRegular struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type Interface interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateRegularChainIfNotExists(table TableType, chain *ChainRegular) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type MssClampInterface interface {
+	InitMultiClusterMssClamp() error
+	DelMultiClusterMssClamp() error
+	SetMultiClusterMssClamp(clampType MssClampType, mssValue int) error
+	AddSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+	DelSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+}
+type GlobalNetEgressInterface interface {
+	CreateGlobalEgressForPods(key, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForPods(handle GlobalEgressHandle) error
+	CreateGlobalEgressForNamespace(namespace, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	UnInstallGlobalEgress() error
+}
+
+type Driver interface {
```

Comment:
I'm not sure of the purpose of the `Driver` interface. We really only need one publicly consumed interface. In the existing iptables API, we define a `Basic` interface that is implemented by the iptables wrapper and the fake one. We also define the public `Interface` that defines higher-level helper methods that use the low-level `Basic` interface. I think we should follow that same paradigm here, assuming similar higher-level  methods still apply, eg `CreateRegularChainIfNotExists`, `PrependUnique` etc.